1-building-an-instagram-hashtag-typeahead-in-javascript/post.md

```diff
@@ -0,0 +1,202 @@
+
+`caretIndex` — In order to determine what part of the post caption the user is editing, we need to know where the caret is. so that we know exactly what part of the post caption the user is editing. E.g. if the user’s caret is here: `#hello worl|` , then the caret index would be `11`. We get this information from `event.target.selectionStart`.
+
+The `getActiveHashtag` function will either return the hashtag the user is editing (e.g. `#worl`), or `null` if no hashtag is actively being edited by the user.
+
+With these arguments in mind and this desired output (either a string that is the hashtag, or `null`), let’s build some pseudocode test cases that we can use to determine whether or not a user is editing the post caption! You should be able to do this with any common testing suite. At Tailwind, we do Test-Driven Development, using [tape](https://www.npmjs.com/package/tape), to reduce bugs and ensure that we’ve planned our implementation correctly before we write the actual code.
+
+The event handler is pretty straightforward. It will call the `getActiveHashtag` method and return the resulting hashtag that it finds:
+```js
+// Event handler for when the user presses 
+// a key inside of the post editor
+onKeyPress = event => {
+    const content = event.target.value;
+    const key = event.key;
+    const caretIndex = event.target.selectionStart;
+    return getActiveHashtag(content, key, caretIndex)
+}
+```
+We named our parser function `getActiveHashtag` and we wrote it to accept the following arguments:
+
+**Case 1 — User is creating or editing a hashtag:**
+```js
+const test => {
+    // Arrange
+    const content = 'Hello #worl';
+    const key = 'l';
+    const caretIndex = 11; // End of line
+
+    // Act
+    const r = getActiveHashtag(content, key, caretIndex);
+
+    // Assert
+    test.assert(r, `#worl`);
+};
+```
+**Case 2 — User’s caret is inside of a hashtag but the user is not editing it:**
+```js
+const test => {
+    // Arrange
+    const content = 'Hello #world';
+    const key = 'ArrowLeft';
+    const caretIndex = 11; // Between the `l` and the `d`
+
+    // Act
+    const r = getActiveHashtag(content, key, caretIndex);
+
+    // Assert
+    test.assert(r, null);
+};
+```
+**Case 3 — User’s caret is not focused on a hashtag:**
+```js
+const test => {
+    // Arrange
+    const content = '#Hello worl';
+    const key = 'l';
+    const caretIndex = 11; // End of line
+
+    // Act
+    const r = getActiveHashtag(content, key, caretIndex);
+
+    // Assert
+    test.assert(r, null);
+};
+```
+Now that we know what results `getActiveHashtag` should return, let’s look at the implementation:
+```js
+// Keys that never used in the process of
+// editing a hashtag.
+const nonEditingKeys = [
+   'ArrowLeft',
+   'ArrowRight',
+   'Control',
+   'Shift',
+   // ...etc
+];
+
+// Returns the actively-being-edited hashtag, or null
+// if none is found.  
+const getActiveHashtag = (content, key, caretIndex) => {
```

Comment:
Since we're just at this start of this journey, and I don't have a feel for how much code to include vs. not include, how did you decide what level of detail to include in the code samples?

Reply:
It's a tough balance and frankly I'm still not sure what the correct level of detail is. I think the goal, though this isn't a tutorial, is to provide just enough code to make this reproducible for someone. I dunno, it would almost be easier to take a half-day and make a working minimal GitHub repo, then build a tutorial around it. I may try that in the future.